File: scripts/migrate_visits_add_adminid.py
#!/usr/bin/env python3
"""
Migration script to add adminId to all visits documents in Firestore.
Requires: firebase-admin, google-cloud-firestore
Install: pip3 install firebase-admin google-cloud-firestore
Run: python3 scripts/migrate_visits_add_adminid.py
"""

import json
import os
import sys
from pathlib import Path

try:
    import firebase_admin
    from firebase_admin import credentials, firestore
except ImportError:
    print("❌ Missing required packages. Install with:")
    print("   pip3 install firebase-admin google-cloud-firestore")
    sys.exit(1)


def find_service_account():
    """Find Firebase service account JSON file."""
    possible_paths = [
        Path.home() / ".config/firebase/cement-delivery-tracker-72de2-key.json",
        Path.home() / ".firebase/cement-delivery-tracker-72de2-key.json",
        Path("firebase-key.json"),
        Path(".env.firebase"),
    ]
    
    for path in possible_paths:
        if path.exists():
            return str(path)
    
    # Check environment variable
    if "FIREBASE_SERVICE_ACCOUNT" in os.environ:
        return os.environ["FIREBASE_SERVICE_ACCOUNT"]
    
    return None


def initialize_firebase():
    """Initialize Firebase Admin SDK."""
    # Try to use service account
    service_account_path = find_service_account()
    
    if service_account_path:
        print(f"✓ Using service account: {service_account_path}")
        cred = credentials.Certificate(service_account_path)
    else:
        # Use Application Default Credentials (from gcloud or environment)
        print("ℹ Using Application Default Credentials")
        cred = credentials.ApplicationDefault()
    
    firebase_admin.initialize_app(cred, {
        "projectId": "cement-delivery-tracker-72de2"
    })
    
    return firestore.client()


def migrate_visits(db):
    """Add adminId to all visits documents."""
    print("\nStarting migration: Adding adminId to visits...\n")
    
    try:
        # Get all visits
        visits_ref = db.collection("visits")
        docs = visits_ref.stream()
        
        visits_list = list(docs)
        total = len(visits_list)
        
        print(f"Found {total} visits to process\n")
        
        if total == 0:
            print("✓ No visits found. Migration complete.")
            return
        
        processed = 0
        updated = 0
        errors = 0
        
        for doc in visits_list:
            visit_id = doc.id
            visit_data = doc.to_dict()
            processed += 1
            
            # Skip if adminId already exists
            if "adminId" in visit_data:
                print(f"✓ [{processed}/{total}] Visit {visit_id} already has adminId: {visit_data['adminId']}")
                continue
            
            try:
                # Get employee's adminId
                employee_id = visit_data.get("employeeId")
                if not employee_id:
                    print(f"✗ [{processed}/{total}] Visit {visit_id} has no employeeId")
                    errors += 1
                    continue
                
                employee_doc = db.collection("users").document(employee_id).get()
                
                if not employee_doc.exists:
                    print(f"✗ [{processed}/{total}] Employee {employee_id} not found for visit {visit_id}")
                    errors += 1
                    continue
                
                employee_data = employee_doc.to_dict()
                admin_id = employee_data.get("adminId")
                
                if not admin_id:
                    print(f"✗ [{processed}/{total}] Employee {employee_id} has no adminId for visit {visit_id}")
                    errors += 1
                    continue
                
                # Update visit with adminId
                db.collection("visits").document(visit_id).update({
                    "adminId": admin_id
                })
                
                updated += 1
                print(f"✓ [{processed}/{total}] Visit {visit_id} → adminId: {admin_id}")
                
            except Exception as e:
                print(f"✗ [{processed}/{total}] Error processing visit {visit_id}: {str(e)}")
                errors += 1
        
        # Print summary
        print(f"\n{'='*60}")
        print("Migration Summary:")
        print(f"  Total visits:   {total}")
        print(f"  Updated:        {updated}")
        print(f"  Errors:         {errors}")
        print(f"  Already had ID: {total - updated - errors}")
        print(f"{'='*60}\n")
        
    except Exception as e:
        print(f"❌ Migration failed: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    try:
        db = initialize_firebase()
        migrate_visits(db)
        print("✓ Migration complete!")
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        print("\nTo use this script:")
        print("1. Install dependencies: pip3 install firebase-admin")
        print("2. Set up credentials:")
        print("   - Option A: Download service account key from Firebase Console")
        print("     Place at: ~/.firebase/cement-delivery-tracker-72de2-key.json")
        print("   - Option B: Use gcloud auth application-default login")
        sys.exit(1)

File: scripts/migrate_visits_add_adminid.js
#!/usr/bin/env node

// Migration script to add adminId to all visits documents
// Usage: node scripts/migrate_visits_add_adminid.js

const { initializeApp, cert } = require("firebase-admin/app");
const { getFirestore } = require("firebase-admin/firestore");
const fs = require("fs");
const os = require("os");
const path = require("path");

// Try to get credentials from Firebase CLI or environment
let credentialsPath = process.env.GOOGLE_APPLICATION_CREDENTIALS;

if (!credentialsPath) {
  // Try common Firebase credential locations
  const commonPaths = [
    path.join(
      os.homedir(),
      ".config/gcloud/application_default_credentials.json",
    ),
    path.join(os.homedir(), ".cache/firebase/credentials.json"),
    "firebase-key.json",
    ".env.firebasekey",
  ];

  for (const p of commonPaths) {
    if (fs.existsSync(p)) {
      credentialsPath = p;
      break;
    }
  }
}

let app;
if (credentialsPath && fs.existsSync(credentialsPath)) {
  const serviceAccount = JSON.parse(fs.readFileSync(credentialsPath, "utf8"));
  app = initializeApp({
    credential: cert(serviceAccount),
  });
} else {
  // Use default credentials (Firebase CLI)
  app = initializeApp({
    projectId: "cement-delivery-tracker-72de2",
  });
}

const db = getFirestore(app);
const BATCH_SIZE = 500;

async function migrateVisits() {
  console.log("Starting migration: Adding adminId to visits...\n");

  try {
    // Get all visits
    const visitsSnapshot = await db.collection("visits").get();
    console.log(`Found ${visitsSnapshot.size} visits to process\n`);

    if (visitsSnapshot.empty) {
      console.log("No visits found. Migration complete.");
      process.exit(0);
    }

    let processed = 0;
    let updated = 0;
    let errors = 0;
    let batch = db.batch();

    for (const visitDoc of visitsSnapshot.docs) {
      const visitData = visitDoc.data();

      // Skip if adminId already exists
      if (visitData.adminId) {
        console.log(
          `✓ [${processed + 1}/${visitsSnapshot.size}] Visit ${visitDoc.id} already has adminId: ${visitData.adminId}`,
        );
        processed++;
        continue;
      }

      try {
        // Get the employee's adminId
        const employeeDoc = await db
          .collection("users")
          .doc(visitData.employeeId)
          .get();

        if (!employeeDoc.exists) {
          console.warn(
            `✗ [${processed + 1}/${visitsSnapshot.size}] Employee ${visitData.employeeId} not found for visit ${visitDoc.id}`,
          );
          errors++;
          processed++;
          continue;
        }

        const employeeData = employeeDoc.data();
        const adminId = employeeData.adminId;

        if (!adminId) {
          console.warn(
            `✗ [${processed + 1}/${visitsSnapshot.size}] Employee ${visitData.employeeId} has no adminId for visit ${visitDoc.id}`,
          );
          errors++;
          processed++;
          continue;
        }

        // Add update to batch
        batch.update(visitDoc.ref, { adminId });
        updated++;
        console.log(
          `✓ [${processed + 1}/${visitsSnapshot.size}] Visit ${visitDoc.id} → adminId: ${adminId}`,
        );

        // Commit batch when it reaches the size limit
        if (updated % BATCH_SIZE === 0) {
          await batch.commit();
          console.log(`\n  ✔ Committed batch of ${BATCH_SIZE} updates\n`);
          batch = db.batch();
        }

        processed++;
      } catch (error) {
        console.error(
          `✗ [${processed + 1}/${visitsSnapshot.size}] Error processing visit ${visitDoc.id}:`,
          error.message,
        );
        errors++;
        processed++;
      }
    }

    // Commit remaining updates
    if (updated % BATCH_SIZE !== 0) {
      await batch.commit();
      console.log(
        `\n✔ Committed final batch of ${updated % BATCH_SIZE} updates`,
      );
    }

    console.log(`\n${"=".repeat(60)}`);
    console.log(`Migration Summary:`);
    console.log(`  Total visits:   ${visitsSnapshot.size}`);
    console.log(`  Updated:        ${updated}`);
    console.log(`  Errors:         ${errors}`);
    console.log(`  Already had ID: ${visitsSnapshot.size - updated - errors}`);
    console.log(`${"=".repeat(60)}\n`);

    process.exit(0);
  } catch (error) {
    console.error("Migration failed:", error);
    process.exit(1);
  }
}

migrateVisits();
